### PR TITLE
Add support for Renesas RZ/N1 platform

### PR DESCRIPTION
**Pull request by b49020**
Add support for RZ/N1 platform from Renasas:
- Cortex-A7 based dual core processor.

This platform supports TrustZone based IO register access control, so add corresponding OEM service based implementation.

For more information about this platform, refer [here](https://www.renesas.com/us/en/products/microcontrollers-microprocessors/rz/rzn/rzn1d.html).

### CODE REVIEW

File: core/arch/arm/plat-ls/main.c [hunk 1]
```diff
@@ -105,6 +105,8 @@ void console_init(void)
 	pl011_init(&console_data, CONSOLE_UART_BASE, CONSOLE_UART_CLK_IN_HZ,
 		   CONSOLE_BAUDRATE);
 #else
+	console_data.io_width = IO_WIDTH_U8;
+	console_data.reg_shift = 0;
```
Comment (jforissier) [inline core/arch/arm/plat-ls/main.c:109]: Minor: personally I would not add those, since `console_data` is guaranteed to be zero-initialized here.

File: core/arch/arm/plat-ls/main.c [hunk 2]
```diff
@@ -105,6 +105,8 @@ void console_init(void)
...
 	ns16550_init(&console_data, CONSOLE_UART_BASE);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-ls/main.c:110]: Any reason why `ns16550_init()` isn't changed to take `io_width` and `reg_shift` as parameters instead?

File: core/drivers/ns16550.c [hunk 3]
```diff
@@ -46,30 +45,27 @@
...
+		UART_LSR_THRE) == 0)
 		;
 }
 
 static void ns16550_putc(struct serial_chip *chip, int ch)
 {
-	vaddr_t base = chip_to_base(chip);
+	struct ns16550_data *pd =
+		container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = io_pa_or_va(&pd->base);
```
Comment (jforissier) [inline core/drivers/ns16550.c:63]: Minor: is this really better than `chip_to_base()`? That's an unrelated change anyways, better as a different commit.

**Review by jforissier — COMMENTED** [review]
Comments on "drivers: ns16550: Allow customizable serial IO config".

**Review by jforissier — COMMENTED** [review]
For "core: Allow non-secure context restore in thumb mode":

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: core/arch/arm/plat-rzn1/conf.mk [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+$(call force,CFG_SM_PLATFORM_HANDLER,y)
+$(call force,CFG_TA_FLOAT_SUPPORT,n)
+
+ta-targets = ta_arm32
+
+$(call force,CFG_TZDRAM_START,0x88000000)
+$(call force,CFG_TZDRAM_SIZE,0x00A00000)
+$(call force,CFG_SHMEM_START,0x87C00000)
+$(call force,CFG_SHMEM_SIZE,0x00400000)
+$(call force,CFG_TEE_RAM_VA_SIZE,0x00200000)
```
Comment (rfs613) [inline core/arch/arm/plat-rzn1/conf.mk:23]: These values represent specific memory layout choices on this particular board/RAM combination. It's quite possible for other users of RZ/N1 to want a different layout.

In kernel land, such values would be in DTB rather than in platform driver. Not sure how other OPTEE targets handle this sort of thing.

Comment (jforissier) [reply core/arch/arm/plat-rzn1/conf.mk:23]: Pretty much hardcoded for all platforms. These values would need to come from a secure DT. Such a beast is partially implemented for a special use case for plat-stm32mp1: a secure DT is embedded in the OP-TEE binary, see `CFG_EMBED_DTB`. I don't think the memory ranges are defined in it however.

So for the time being, these values should probably be defined as `CFG_xxx ?= value` so they can be easily changed at build time.

**Review by jforissier — COMMENTED** [review]
Reviewing "Add support for plat-rzn1":

- (Minor) The subject would be more explicit as "Add support for Renesas RZ/N1" (and then you mention `PLATFORM=rzn1` in the description)
- Please add the link to the hardware documentation in a Link: tag (I know it can/will become stale one day but it doesn't hurt to have it until then...)

Minor comments below. In any case:

`Acked-by: Jerome Forissier <jerome@forissier.org>`

File: core/arch/arm/plat-ls/main.c (hunk 2 above)
Comment (b49020) [reply core/arch/arm/plat-ls/main.c:110]: I only saw scalability being the reason but since now we only have 3 extra params apart from `struct console_data`, so I can pass them as parameters as well.

File: core/arch/arm/plat-ls/main.c (hunk 1 above)
Comment (b49020) [reply core/arch/arm/plat-ls/main.c:109]: As per @jenswi-linaro comment, I will rather pass them as params to `ns16550_init()`.

File: core/drivers/ns16550.c (hunk 3 above)
Comment (b49020) [reply core/drivers/ns16550.c:63]: `chip_to_base()` only allowed us to retrieve `base` value but now we need to retrieve `reg_shift` and `io_width` values as well. So in order to avoid duplicate code, I removed `chip_to_base()`.

File: core/arch/arm/plat-rzn1/conf.mk (hunk 4 above)
Comment (b49020) [reply core/arch/arm/plat-rzn1/conf.mk:23]: @rfs613

> These values represent specific memory layout choices on this particular board/RAM combination. It's quite possible for other users of RZ/N1 to want a different layout.

Yes, so to handle this scenario I would prefer to add a new `PLATFORM_FLAVOR`. So based on that platform flavour we can define the memory layout.

@jforissier 

> a secure DT is embedded in the OP-TEE binary, see CFG_EMBED_DTB. I don't think the memory ranges are defined in it however.

I think we should be able to retrieve TA memory and shared memory info from a secure DT. But I am not sure how we can make `TEE_LOAD_ADDR` (fixed based address at compile time) configurable via DT. Probably to address this we need to add Position Independent Executable (PIE) feature to OP-TEE OS.

> So for the time being, these values should probably be defined as CFG_xxx ?= value so they can be easily changed at build time.

Sounds reasonable to me, will use it instead.

File: core/arch/arm/plat-rzn1/a7_plat_init.S [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+.code 32
+
+/*
+ * platform early configuration
+ *
+ * Use scratables registers R0-R3.
+ * No stack usage.
+ * LR store return address.
+ * Trap CPU in case of error.
+ */
```
[the remark this answers is not in the document]
Comment (b49020) [reply core/arch/arm/plat-rzn1/a7_plat_init.S:33]: Okay.

File: core/arch/arm/plat-rzn1/a7_plat_init.S [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	 * - FIQ: In Sec world, FIQ trapped to FIQ Mode [bit2=0]
+	 * - IRQ: IRQ always trapped to IRQ Mode [bit1=0]
+	 * - NS: Secure World [bit0=0]
+	 */
+	mov r0, #SCR_AW
+	write_scr r0
+
+	/*
+	 * SCTLR = 0x00000000
+	 */
```
[the remark this answers is not in the document]
Comment (b49020) [reply core/arch/arm/plat-rzn1/a7_plat_init.S:51]: Okay.

**Comment by b49020** [comment]
> (Minor) The subject would be more explicit as "Add support for Renesas RZ/N1" (and then you mention PLATFORM=rzn1 in the description)

Okay.

> Please add the link to the hardware documentation in a Link: tag (I know it can/will become stale one day but it doesn't hurt to have it until then...)

Sure.

File: core/arch/arm/plat-rzn1/conf.mk (hunk 4 above)
Comment (jforissier) [reply core/arch/arm/plat-rzn1/conf.mk:23]: Agreed. We have `-fpie` already, with ASLR.

Comment (b49020) [reply core/arch/arm/plat-rzn1/conf.mk:23]: > We have -fpie already, with ASLR.

Okay, I see. So I think it should be doable to have position independent loading of `tee.bin`.

**Comment by b49020** [comment]
Addressed comments and tags applied.

File: core/drivers/ns16550.c (hunk 3 above)
Comment (jforissier) [reply core/drivers/ns16550.c:63]: Ah yes I see. That's fine.

**Comment by jforissier** [comment]
For "drivers: ns16550: Allow customizable serial IO config":

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by b49020** [comment]
Thanks @jforissier for your review, tag applied.

File: core/arch/arm/plat-rzn1/main.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+void console_init(void)
+{
+	ns16550_init(&console_data, CONSOLE_UART_BASE, IO_WIDTH_U32, 2);
+	register_serial_console(&console_data.chip);
+}
+
+void main_init_gic(void)
+{
+	vaddr_t gicc_base;
+	vaddr_t gicd_base;
```
Comment (etienne-lms) [inline core/arch/arm/plat-rzn1/main.c:35]: minor: prefer with an initialization value, for consistency and [coding standards](https://optee.readthedocs.io/en/latest/general/coding_standards.html#coding-standards).
Ditto in `rzn1_tz_init()`.

File: core/arch/arm/plat-rzn1/sm_platform_handler.c [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+static uint32_t oem_sysreg(uint32_t addr, uint32_t mask, uint32_t *pvalue)
+{
+	vaddr_t reg = 0;
+	const struct regauth_t *auth = get_regauth(addr);
+
+	/* Allow operations on registers not in the list */
+	if (!auth)
+		auth = &regauth_pass;
+
+	reg = core_mmu_get_va(addr, MEM_AREA_IO_SEC);
```
[the remark this answers is not in the document]
Comment (b49020) [reply core/arch/arm/plat-rzn1/sm_platform_handler.c:44]: Error code isn't checked in Linux for normal register read/write. So will rather add check for `!reg` and corresponding `DMSG`.

File: core/arch/arm/plat-rzn1/sm_platform_handler.c [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		mask &= auth->wmask;
+		if (mask == 0U)
+			DMSG("Blocking write of 0x%"PRIx32" to register 0x%"
+			     PRIx32" (0x%"PRIxVA")", *pvalue, addr, reg);
+		else if (mask == ~0U)
+			io_write32(reg, *pvalue);
+		else
+			/* Partial write with read ahead */
+			io_write32(reg, (io_read32(reg) & ~mask) |
+					(*pvalue & mask));
```
[the remark this answers is not in the document]
Comment (b49020) [reply core/arch/arm/plat-rzn1/sm_platform_handler.c:57]: Okay, will use it instead.

File: core/arch/arm/plat-rzn1/main.c (hunk 7 above)
Comment (b49020) [reply core/arch/arm/plat-rzn1/main.c:35]: Okay.

**Comment by b49020** [comment]
Addressed comments.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for all commits, with late minor comments addressed or not.

File: core/arch/arm/plat-rzn1/psci.c [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	vaddr_t ctrl_reg = 0;
+	uint32_t value = 0;
+
+	en_reg = core_mmu_get_va(SYSCTRL_REG_RSTEN, MEM_AREA_IO_SEC);
+	ctrl_reg = core_mmu_get_va(SYSCTRL_REG_RSTCTRL, MEM_AREA_IO_SEC);
+
+	/* Enable software reset */
+	value = io_read32(en_reg);
+	value |= SYSCTRL_REG_RSTEN_SWRST_EN | SYSCTRL_REG_RSTEN_MRESET_EN;
+	io_write32(en_reg, value);
```
[the remark this answers is not in the document]
Comment (b49020) [reply core/arch/arm/plat-rzn1/psci.c:86]: Looks cleaner, will use it instead.

File: core/arch/arm/plat-rzn1/sm_platform_handler.c [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+static const struct regauth_t regauth_pass = {.rmask = ~0U, .wmask = ~0U};
+
+static const struct regauth_t *get_regauth(unsigned long paddr)
+{
+	unsigned int idx = 0;
+	unsigned int len = ARRAY_SIZE(regauth);
+
+	while (idx < len) {
+		if (core_is_buffer_inside(paddr, 0x4, regauth[idx].paddr,
```
[the remark this answers is not in the document]
Comment (b49020) [reply core/arch/arm/plat-rzn1/sm_platform_handler.c:26]: Okay.

**Comment by b49020** [comment]
Thanks @etienne-lms for your review, squashed changes and tag applied.
